gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua

```diff
@@ -0,0 +1,150 @@
+local backend_client = require('apicast.backend_client')
+local AuthsCache = require('auths_cache')
+local ReportsBatcher = require('reports_batcher')
+local policy = require('apicast.policy')
+local errors = require('apicast.errors')
+local reporter = require('reporter')
+local http_ng_resty = require('resty.http_ng.backend.resty')
+local semaphore = require('ngx.semaphore')
+
+local ipairs = ipairs
+
+local default_auths_ttl = 10
+local default_batch_reports_seconds = 10
+
+local _M = policy.new('Caching policy')
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new(config)
+
+  local auths_ttl = config.auths_ttl or default_auths_ttl
+  self.auths_cache = AuthsCache.new(ngx.shared.cached_auths, auths_ttl)
+
+  self.reports_batcher = ReportsBatcher.new(
+    ngx.shared.batched_reports, 'batched_reports_locks')
+
+  self.batch_reports_seconds = config.batch_reports_seconds or
+                               default_batch_reports_seconds
+
+  self.report_timer_on = false
+
+  -- Semaphore used to ensure that only one timer is started per worker.
+  local semaphore_report_timer, err = semaphore.new(1)
+  if not semaphore_report_timer then
+    ngx.log(ngx.ERR, "Create semaphore failed: ", err)
+  end
+  self.semaphore_report_timer = semaphore_report_timer
+
+  return self
+end
+
+-- TODO: More policies are using this method. Move it to backend_client to
+-- avoid duplicating code.
+-- Converts a usage to the format expected by the 3scale backend client.
+local function format_usage(usage)
+  local res = {}
+
+  local usage_metrics = usage.metrics
+  local usage_deltas = usage.deltas
+
+  for _, metric in ipairs(usage_metrics) do
+    local delta = usage_deltas[metric]
+    res['usage[' .. metric .. ']'] = delta
```

Comment:
No need to change this now, but this should also use `string.format` to avoid the allocations.